common.c

```diff
@@ -1987,3 +1961,143 @@ static int rtr_check_logging_config(struct rtr_event_info *event_info, int stack
 
 	return 1;
 }
+
+/* get net fuzzing flag */
+static const char *net_func_names[] = {
+	"socket", "connect", "bind", "listen", "accept", "send",
+	"recv", "sendto", "recvfrom", "sendmsg", "recvmsg",
+	"gethostbyname", "gethostbyaddr", "getaddrinfo", NULL
+};
+
+static const char *net_fuzz_types[] = {
+	"NO_MEMORY", "LIMIT_SOCKET", "ADDR_INUSE", "NET_UNREACHABLE",
+	"CONN_TIMEOUT", "CONN_RESET", "CONN_REFUSE",
+	"HOST_NOT_FOUND", "SERVICE_NOT_AVAIL",
+	NULL
+};
+
+static const int net_fuzz_errnos[] = {
+	ENOMEM, ENFILE, EADDRINUSE, ENETUNREACH, ETIMEDOUT,
+	ECONNRESET, ECONNREFUSED, HOST_NOT_FOUND, TRY_AGAIN,
+};
+
+static void init_netfuzz_config(void)
+{
+	RTR_CONFIG_HANDLE config = NULL;
+
+	while (1) {
+		char *func_name = NULL;
+		char *fuzz_type_str = NULL;
+		double fuzz_rate;
+
+		int i, func_idx = -1, fuzz_type = -1;
+
+		if (rtr_get_config_multiple(&config, "fuzzing-net", ARGUMENT_TYPE_STRING, ARGUMENT_TYPE_STRING,
+			ARGUMENT_TYPE_DOUBLE, ARGUMENT_TYPE_END, &func_name, &fuzz_type_str, &fuzz_rate) == 0)
+			break;
+
+		/* get func index from func name */
+		for (i = 0; net_func_names[i] != NULL; i++) {
+			if (real_strcmp(func_name, net_func_names[i]) == 0) {
+				func_idx = i;
+				break;
+			}
+		}
+
+		/* get fuzzing type */
+		for (i = 0; net_fuzz_types[i] != NULL; i++) {
+			if (real_strcmp(fuzz_type_str, net_fuzz_types[i]) == 0) {
+				fuzz_type = (i == 0) ? NET_FUZZ_TYPE_NOMEM : i * 2;
+				break;
+			}
+		}
+
+		/* set fuzzing value */
+		if (func_idx >= 0 && fuzz_type >= 0) {
+			g_netfuzz_config.fuzz_types[func_idx] = fuzz_type;
+			g_netfuzz_config.fuzz_rates[func_idx] = fuzz_rate;
+		}
+
+		if (!config)
+			break;
+	}
+
+	g_netfuzz_config.init_flag = 1;
+}
+
+int rtr_get_net_fuzzing(const char *func_name, int *err_val)
+{
+	int func_idx = -1;
+	int fuzz_typeset, fuzz_type;
+
+	char *func, *fuzz_type_str;
+	int fail_chance = 0;
+
+	int matched = 0;
+
+	/* init fuzzing config */
+	if (!g_netfuzz_config.init_flag)
+		init_netfuzz_config();
+
+	/* get index of socket function */
+	for (func_idx = 0; net_func_names[func_idx] != NULL; func_idx++) {
+		if (real_strcmp(func_name, net_func_names[func_idx]) == 0)
+			break;
+	}
+
+	/* check matched status */
+	if (func_idx < 0)
```

Comment:
I'm not sure this can ever happen :) You mean `if (net_func_names[func_idx] == NULL)` Also `net_func_names` and `enum RTR_NET_FUNCTION_IDX` could easily get out of sync. Perhaps something like:

```
struct {
    enum RTR_NET_FUNC_ID func_id ;
    const char *name;
} net_funcs = [{
    {NET_FUNC_SOCKET, "socket"},
    {NET_FUNC_CONNECT, "connect"},
    ...
    {NET_FUNC_END, NULL}
};

enum RTR_NET_FUNC_ID func_id = NET_FUNC_END;
int i;
for (i=0; net_funcs[i].func_id != NET_FUNC_END; i++) {
    if (real_strcmp(func_name, net_funcs[i].name) == 0) {
        func_id = net_funcs[i].id;
        break;
    }
}
if (func_id != RTR_NET_FUNC_END)
    return 0;
```